sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -394,83 +410,349 @@ struct codec_info codec_info_list[] = {
+						       class_id);
+		else
+			codec[i].name = devm_kasprintf(dev, GFP_KERNEL,
+						       "sdw:%x:%x:%x:%x:%x",
+						       link_id, mfg_id, part_id,
+						       class_id, unique_id);
+
+		if (!codec[i].name)
+			return -ENOMEM;
+
+		id = find_codec_info(part_id);
+		if (id < 0)
+			return id;
+
+		codec[i].dai_name = codec_info_list[id].dai_name;
+	}
+
+	return 0;
+}
+
+static void set_codec_init_func(const struct snd_soc_acpi_link_adr *link,
+				struct snd_soc_dai_link *dai_links,
+				bool playback)
+{
+	unsigned int part_id;
+	int i, id;
+
+	for (i = 0; i < link->num_adr; i++) {
+		part_id = (link->adr[i] >> 8) & GENMASK(15, 0);
+		id = find_codec_info(part_id);
+
+		if (codec_info_list[id].init)
+			codec_info_list[id].init(link, dai_links, playback);
+	}
+}
+
+static int create_sdw_codec_dai(struct device *dev,
+				struct snd_soc_dai_link *dai_links, int id,
+				struct snd_soc_dai_link_component *cpus,
+				const struct snd_soc_acpi_link_adr *link)
+{
+	struct snd_soc_dai_link_component *codec;
+	unsigned int part_id, link_id;
+	int i, j = 0, idx;
+	int ret;
+
+	codec = devm_kcalloc(dev, link->num_adr,
+			     sizeof(struct snd_soc_dai_link_component),
+			     GFP_KERNEL);
+	if (!codec)
+		return -ENOMEM;
+
+	ret = create_codec_dai_name(dev, link, codec);
+	if (ret < 0)
+		return ret;
+
+	part_id = (link->adr[0] >> 8) & GENMASK(15, 0);
+	idx = find_codec_info(part_id);
+	if (idx < 0)
+		return idx;
+
+	link_id = ffs(link->mask) - 1;
+
+	/* playback & capture */
+	for (i = 0; i < 2; i++) {
+		char *name, *cpu_name;
+
+		if (!codec_info_list[idx].direction[i])
+			continue;
+
+		name = devm_kasprintf(dev, GFP_KERNEL,
+				      sdw_stream_name[i], link_id);
+		if (!name)
+			return -ENOMEM;
+
+		cpu_name = devm_kasprintf(dev, GFP_KERNEL,
+					  "SDW%d Pin%d", link_id, j + 2);
+		if (!cpu_name)
+			return -ENOMEM;
+
+		init_dai_link(dai_links, id + j, name, cpus, cpu_name,
+			      codec, link->num_adr, 1 - i, i);
+		set_codec_init_func(link, dai_links + id + j, 1 - i);
+		j++;
+	}
+
+	return j;
+}
+
+static int sof_card_dai_links_create(struct device *dev,
+				     struct snd_soc_acpi_mach *mach,
+				     struct snd_soc_card *card)
+{
+	struct snd_soc_dai_link_component *idisp_components;
+	struct snd_soc_acpi_mach_params *mach_params;
+	int ssp_num = 0, sdw_num, hdmi_num = 0;
+	const struct snd_soc_acpi_link_adr *adr_link;
+	struct snd_soc_dai_link_component *cpus;
+	struct snd_soc_dai_link *links;
+	int num_links;
+	int i, id = 0;
+	int val;
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA_COMMON_HDMI_CODEC)
+	hdmi_num = sof_rt711_rt1308_rt715_quirk & SOF_SDW_TGL_HDMI ?
+				SOF_TGL_HDMI_COUNT : SOF_PRE_TGL_HDMI_COUNT;
 #endif
-	{
-		.name = "SDW2-Playback",
-		.id = 3,
-		.init = second_spk_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		.nonatomic = true,
-		SND_SOC_DAILINK_REG(sdw2_pin2, sdw2_codec, platform),
-	},
-};
+
+	/* check ACPI id for I2S codec */
+	if (acpi_dev_present(mach->id, NULL, -1))
```

Comment:
you should first test if mach->id is set

Reply:
done